File: rust/src/api/alerts_api.rs
use std::fmt::Debug;
use std::future::Future;
use std::pin::Pin;
use std::sync::Arc;
use futures_util::future;
use log::debug;
use reqwest::{RequestBuilder, Response, Url};
use tokio::sync::Semaphore;
use crate::data::httpapi::{NotifyRequest, QueryAlertsResponse, QueryResponse, ResolveRequest};
use crate::data::httpapi::alerts::QueryName;

#[derive(Clone)]
pub struct BaseUri(pub Url);

pub struct MaxConcurrentRequests(pub usize);

#[derive(Clone)]
pub struct Client {
    base_uri: BaseUri,
    max_concurrent_requests: Option<Arc<Semaphore>>,
    client: reqwest::Client
}

impl Client {
    pub fn new(base_uri: BaseUri, max_concurrent_requests: Option<MaxConcurrentRequests>) -> Client {
        let max_concurrent_requests =
            max_concurrent_requests.map(|max|
                Arc::new(Semaphore::new(max.0))
            );
        Client { base_uri, max_concurrent_requests, client: reqwest::Client::new() }
    }

    pub fn query_alerts(&self) -> impl Future<Output = reqwest::Result<QueryAlertsResponse>> {
        let url = self.base_uri.0.join("alerts").unwrap();
        let request = self.client.get(url);
        self.do_request(
            request,
            |response| Box::pin(response.json::<QueryAlertsResponse>())
        )
    }

    pub fn query(
        &self, query_name: &QueryName
    ) -> impl Future<Output = reqwest::Result<QueryResponse>> {
        let mut url = self.base_uri.0.join("query").unwrap();
        url.query_pairs_mut().append_pair("target", query_name.0.as_str());
        let request = self.client.get(url);
        self.do_request(
            request,
            |response| Box::pin(response.json::<QueryResponse>())
        )
    }

    pub fn notify(
        &self, data: &NotifyRequest
    ) -> impl Future<Output = reqwest::Result<()>> {
        let url = self.base_uri.0.join("notify").unwrap();
        let request = self.client.post(url).json(data);
        self.do_request(
            request,
            |response| Box::pin(future::ready(
                response.error_for_status().map(|_| ())
            ))
        )
    }

    pub fn resolve(
        &self, data: &ResolveRequest
    ) -> impl Future<Output = reqwest::Result<()>> {
        let url = self.base_uri.0.join("resolve").unwrap();
        let request = self.client.post(url).json(data);
        self.do_request(
            request,
            |response| Box::pin(future::ready(
                response.error_for_status().map(|_| ())
            ))
        )
    }

    fn do_request<Parsed : Debug>(
        &self, request: RequestBuilder,
        parse: impl FnOnce(Response) -> Pin<Box<dyn Future<Output = reqwest::Result<Parsed>> + Send>>
    ) -> impl Future<Output = reqwest::Result<Parsed>> {
        let maybe_semaphore = self.max_concurrent_requests.clone();
        // Untie lifetime of the returned future from `self`.
        async move {
            let _permit;
            if let Some(semaphore) = &maybe_semaphore {
                debug!("Getting a permit...");
                _permit = semaphore.acquire().await.unwrap();
                debug!("Permit acquired, sending request.");
            }

            let response = request.send().await?;
            let parsed = parse(response).await;
            debug!("Response received: {:?}", parsed);
            parsed
        }
    }
}

File: go/api/alerts_api.go
package api

import (
	"alert-execution-engine/data/httpapi"
	. "alert-execution-engine/functional"
	"context"
	"encoding/json"
	"fmt"
	"golang.org/x/sync/semaphore"
	"net/http"
	"net/url"
)
import HttpClient "github.com/bozd4g/go-http-client"

type Client struct {
	BaseUri                 url.URL
	maybeConcurrentRequests *semaphore.Weighted
	client                  HttpClient.Client
}

type MaxConcurrentRequests uint

func NewClient(baseUri url.URL, maxConcurrentRequests Option[MaxConcurrentRequests]) Client {
	var concurrentRequests *semaphore.Weighted
	if maxConcurrentRequests.IsSome {
		concurrentRequests = semaphore.NewWeighted(int64(maxConcurrentRequests.UnsafeValue))
	}

	return Client{
		BaseUri:                 baseUri,
		maybeConcurrentRequests: concurrentRequests,
		client:                  HttpClient.New(baseUri.String()),
	}
}

func (self *Client) QueryAlerts() (*httpapi.QueryAlertsResponse, error) {
	request, err := self.client.Get("/alerts")
	if err != nil {
		return nil, err
	}

	arr, err := sendAndParse[[]httpapi.Alert](self, request)
	if err != nil {
		return nil, err
	}

	return &httpapi.QueryAlertsResponse{Alerts: *arr}, nil
}

func (self *Client) Query(queryName httpapi.QueryName) (*httpapi.QueryResponse, error) {
	queryParams := struct {
		Target string `url:"target"`
	}{Target: queryName.String()}
	request, err := self.client.GetWith("/query", queryParams)
	if err != nil {
		return nil, err
	}

	return sendAndParse[httpapi.QueryResponse](self, request)
}

func (self *Client) Notify(requestData httpapi.NotifyRequest) error {
	request, err := self.client.PostWith("/notify", requestData)
	if err != nil {
		return err
	}

	_, err = self.sendAndFailOnNonSuccess(request)
	return err
}

func (self *Client) Resolve(requestData httpapi.ResolveRequest) error {
	request, err := self.client.PostWith("/resolve", requestData)
	if err != nil {
		return err
	}

	_, err = self.sendAndFailOnNonSuccess(request)
	return err
}

type BodyParsingError struct {
	Body         []byte
	ParsingError error
}

func (self BodyParsingError) Error() string {
	return fmt.Sprintf("Cannot parse JSON: %s\nJSON:\n%s", self.ParsingError, string(self.Body))
}

func sendAndParse[Parsed any](self *Client, request *http.Request) (*Parsed, error) {
	response, err := self.send(request)
	if err != nil {
		return nil, err
	}

	body := response.Get().Body
	var parsed Parsed
	err = json.Unmarshal(body, &parsed)
	if err != nil {
		return nil, &BodyParsingError{
			Body:         body,
			ParsingError: err,
		}
	}

	return &parsed, nil
}

type NonSuccessfulRequestError struct {
	Response HttpClient.Response
}

func (self NonSuccessfulRequestError) Error() string {
	return fmt.Sprintf("Response status code %d is not in the 2xx range", self.Response.Get().StatusCode)
}

// Sends the request, fails if the response does not have 2xx status code.
func (self *Client) sendAndFailOnNonSuccess(request *http.Request) (HttpClient.Response, error) {
	response, err := self.send(request)
	if err != nil {
		return nil, err
	}

	statusCode := response.Get().StatusCode
	if statusCode < 200 || statusCode > 299 {
		return nil, NonSuccessfulRequestError{Response: response}
	}

	return response, nil
}

func (self *Client) send(request *http.Request) (HttpClient.Response, error) {
	if self.maybeConcurrentRequests != nil {
		ctx := context.TODO()
		err := self.maybeConcurrentRequests.Acquire(ctx, 1)
		if err != nil {
			return nil, err
		}
		defer self.maybeConcurrentRequests.Release(1)
	}
	return self.client.Do(request)
}
